network/gossip/libp2p/nodeIDRefresher.go

```diff
@@ -0,0 +1,74 @@
+package libp2p
+
+import (
+	"fmt"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/flow/filter"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/state/protocol/events"
+)
+
+// NodeIDRefresher derives the latest list of flow Identities with which the network should be communicating based on the
+// current epoch phase
+type NodeIDRefresher struct {
+	events.Noop
+	logger   zerolog.Logger
+	state    protocol.ReadOnlyState
+	callBack func(flow.IdentityList) error // callBack to call when the id list has changed
+}
+
+func NewNodeIDRefresher(logger zerolog.Logger, state protocol.ReadOnlyState, callBack func(list flow.IdentityList) error) *NodeIDRefresher {
+	return &NodeIDRefresher{
+		logger:   logger,
+		state:    state,
+		callBack: callBack,
+	}
+}
+
+func (listener *NodeIDRefresher) EpochSetupPhaseStarted(newEpoch uint64, first *flow.Header) {
+	log := listener.logger.With().Uint64("new_epoch", newEpoch).Logger()
+
+	// get the new set of IDs
+	newIDs, err := IDsFromState(listener.state)
+	if err != nil {
+		log.Err(err).Msg("failed to update network ids on epoch transition")
+		return
+	}
+
+	// call the registered callback
+	err = listener.callBack(newIDs)
+	if err != nil {
+		log.Err(err).Msg("failed to update network ids on epoch transition")
+		return
+	}
+}
+
+// IDsFromState returns the identities that the network should be using based on the current epoch phase
```

Comment:
I think it is worth adding a note here about how this works and the limitations, because it's a bit non-obvious.

For example:
* it will currently omit nodes from previous epoch if we restart during `StakingPhase`
* it includes nodes from previous epoch otherwise _because_ we wait to call this function until the transition to setup phase, before that transition we retain what was set during the previous epoch